docs: organize cli docs & add page 'integrate-with-gitlab' (#193)
bytebase/bytebase.com: content/docs/en/cli/integrate-with-gitlab.md

---

> title: Integrating with GitLab

Reviewer: ```suggestion
title: Integrating with GitLab CI
```

description: This page contains a tutorial about integrating Bytebase CLI bb with GitLab CI/CD pipeline.
---

> This tutorial will show you the steps to integrate Bytebase CLI bb into your GitLab CI/CD pipeline. You will set up a pipeline automating the database schema change along with the code change, which brings your team’s DevOps automation to a new level.

Reviewer: ```suggestion
This tutorial will show you the steps to integrate Bytebase CLI `bb` into your GitLab CI/CD pipeline. You will set up a pipeline automating the database schema change along with the code change, which brings your team’s DevOps automation to a new level.
```

> This tutorial will show you the steps to integrate Bytebase CLI `bb` into your GitLab CI/CD pipeline. You will set up a pipeline automating the database schema change along with the code change, which brings your team’s DevOps automation to a new level.

Reviewer: Since we have a sample project. It's always desirable to include the link to that project in the summary section.

## Prerequisites

- A GitLab Project. (GitLab.com or GitLab CE/EE. We use GitLab.com for this tutorial.)

> - Self-managed GitLab Runner on a Linux machine. (If haven’t, you can follow the [GitLab Runner docs](https://docs.gitlab.com/runner) to install)

Reviewer: ```suggestion
- Self-managed GitLab Runner on a Linux machine. (You can follow the [GitLab Runner docs](https://docs.gitlab.com/runner) to install)
```

> - A MySQL database that is accessible by the GitLab Runner.

Reviewer: ```suggestion
- A MySQL database that is network accessible by the GitLab Runner.
```

## Install `bb` on GitLab Runner

On your GitLab Runner shell prompt, paste the following command:

```bash
/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/bytebase/bytebase/HEAD/scripts/install_bb.sh)"
```

Run `bb --version` to check if `bb` is installed.

You can go back to [overview](https://www.bytebase.com/docs/cli/overview) for more installation details.

## Register GitLab Runner to a GitLab Project

> In your GitLab Project, go to **Settings > CI/CD > Runners** and follow the **_Show Runner Installation Instructions_** to register runner.

Reviewer: ```suggestion
In your GitLab Project, go to **Settings > CI/CD > Runners** and click the **_Show Runner Installation Instructions_** button to register runner.
```

![register gitlab runner](/static/docs-assets/gitlab-register-runner.png)

> In register interaction, add tag `bb-runner` to specify this runner has `bb` installed.

Reviewer: ```suggestion
In the shell prompt, add tag `bb-runner` to specify this runner has `bb` installed.
```

![tag for gitlab runner](/static/docs-assets/tag-gitlab-runner.png)

## Configure GitLab CI

Set up `.gitlab-ci.yml`:

```yaml
stages:
  - deploy

execute-schema-migration:
  tags: ["bb-runner"] # Runs this job on runners that has bb installed
  only: # Only trigger when a script added to dev/
    refs: ["main"]
    changes: ["dev/*"]
  stage: deploy
  script:
    - bb dump --dsn $BYTEBASE_DSN --schema-only --file before-migrate.sql # Snapshot the schema before migration
    - migrate_scripts=$(git diff-tree --no-commit-id --name-only -r $CI_COMMIT_SHA | grep '^dev/' -) # Extract the added files
    - |
      for migrate_script in $migrate_scripts; do
        bb migrate --dsn $BYTEBASE_DSN --file $migrate_script;
      done
    - bb dump --dsn $BYTEBASE_DSN --schema-only --file after-migrate.sql # Snapshot the schema after migration

  artifacts:
    paths:
      - before-migrate.sql
      - after-migrate.sql
    expire_in: 30 days
```

> ## Add DSN Variable

Reviewer: ```suggestion
## Configure DSN
```

> bb uses DSNs to access databases. Here are a few examples:

Reviewer: ```suggestion
`bb` uses DSNs to access databases. Here are a few examples:
```

> `bb` uses DSNs to access databases. Here are a few examples:

Reviewer: Add the link to the DSN doc

## Prerequisites

- A GitLab Project. (GitLab.com or GitLab CE/EE. We use GitLab.com for this tutorial.)
- Self-managed GitLab Runner on a Linux machine. (You can follow the [GitLab Runner docs](https://docs.gitlab.com/runner) to install)
- A MySQL database that is network accessible by the GitLab Runner.

## Install `bb` on GitLab Runner

On your GitLab Runner shell prompt, paste the following command:

```bash
/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/bytebase/bytebase/HEAD/scripts/install_bb.sh)"
```

Run `bb --version` to check if `bb` is installed.

You can go back to [overview](https://www.bytebase.com/docs/cli/overview) for more installation details.

## Register GitLab Runner to a GitLab Project

In your GitLab Project, go to **Settings > CI/CD > Runners**, click the **_Show Runner Installation Instructions_** button, and follow the instructions to register runner.

![register gitlab runner](/static/docs-assets/gitlab-register-runner.png)

> In the shell prompt, add tag `bb-runner` to specify this runner has `bb` installed.

Reviewer: ```suggestion
From the shell prompt, add tag `bb-runner` to specify this runner has `bb` installed.
```

![tag for gitlab runner](/static/docs-assets/tag-gitlab-runner.png)

## Configure GitLab CI

Set up `.gitlab-ci.yml`:

```yaml
stages:
  - deploy

execute-schema-migration:
  tags: ["bb-runner"] # Runs this job on runners that has bb installed
  only: # Only trigger when a script added to dev/
    refs: ["main"]
    changes: ["dev/*"]
  stage: deploy
  script:
    - bb dump --dsn $BYTEBASE_DSN --schema-only --file before-migrate.sql # Snapshot the schema before migration
    - migrate_scripts=$(git diff-tree --no-commit-id --name-only -r $CI_COMMIT_SHA | grep '^dev/' -) # Extract the added files
    - |
      for migrate_script in $migrate_scripts; do
        bb migrate --dsn $BYTEBASE_DSN --file $migrate_script;
      done
    - bb dump --dsn $BYTEBASE_DSN --schema-only --file after-migrate.sql # Snapshot the schema after migration

  artifacts:
    paths:
      - before-migrate.sql
      - after-migrate.sql
    expire_in: 30 days
```

## Configure DSN

> `bb` uses [DSN](./reference#data-source-name-dsn)s to access databases. Here are a few examples:

Reviewer: ```suggestion
`bb` configures [--dsn](./reference#data-source-name-dsn) to access databases. Here are a few DSN examples:
```

- mysql://root@localhost:3306/
- postgresql://user:pass@localhost:5432/dbname?ssl-ca=a&ssl-cert=b&ssl-key=c

You also need to setup DSN secret to make this workflow available:
Go to **Setting > CI/CD > Variables**, add your DSN as `BYTEBASE_DSN`.

![add dsn secret](/static/docs-assets/add-gitlab-secret-1.png)

![add dsn secret](/static/docs-assets/add-gitlab-secret-2.png)

> ## Trigger a migration

Reviewer: ```suggestion
## Trigger a Migration
```
